File: app/store/state.go
package store

import (
	"database/sql"
	"errors"
	"time"
)

// Per-workspace state (lenses, merges, facets, arrangements) and global
// settings. Values are opaque JSON strings owned by the frontend; the store
// only keeps them durable. They lived in the webview's localStorage, which a
// reinstall, a new dev origin or a cleared cache silently empties.

// GetState returns every key stored for a workspace.
func (s *Store) GetState(workspaceID string) (map[string]string, error) {
	rows, err := s.db.Query(`SELECT key, value FROM workspace_state WHERE workspace_id = ?`, workspaceID)
	if err != nil {
		return nil, err
	}
	defer rows.Close()
	out := map[string]string{}
	for rows.Next() {
		var k, v string
		if err := rows.Scan(&k, &v); err != nil {
			return nil, err
		}
		out[k] = v
	}
	return out, rows.Err()
}

// PutState writes one key; an empty value deletes it.
func (s *Store) PutState(workspaceID, key, value string) error {
	if value == "" {
		_, err := s.db.Exec(`DELETE FROM workspace_state WHERE workspace_id = ? AND key = ?`, workspaceID, key)
		return err
	}
	_, err := s.db.Exec(`
INSERT INTO workspace_state (workspace_id, key, value, updated_at) VALUES (?, ?, ?, ?)
ON CONFLICT (workspace_id, key) DO UPDATE SET value = excluded.value, updated_at = excluded.updated_at`,
		workspaceID, key, value, time.Now().UTC())
	return err
}

// GetSetting returns a global setting, or "" when unset.
func (s *Store) GetSetting(key string) (string, error) {
	var v string
	err := s.db.QueryRow(`SELECT value FROM settings WHERE key = ?`, key).Scan(&v)
	if errors.Is(err, sql.ErrNoRows) {
		return "", nil
	}
	return v, err
}

// PutSetting writes a global setting; an empty value deletes it.
func (s *Store) PutSetting(key, value string) error {
	if value == "" {
		_, err := s.db.Exec(`DELETE FROM settings WHERE key = ?`, key)
		return err
	}
	_, err := s.db.Exec(`INSERT INTO settings (key, value) VALUES (?, ?) ON CONFLICT (key) DO UPDATE SET value = excluded.value`, key, value)
	return err
}

// GetSettings returns every global setting.
func (s *Store) GetSettings() (map[string]string, error) {
	rows, err := s.db.Query(`SELECT key, value FROM settings`)
	if err != nil {
		return nil, err
	}
	defer rows.Close()
	out := map[string]string{}
	for rows.Next() {
		var k, v string
		if err := rows.Scan(&k, &v); err != nil {
			return nil, err
		}
		out[k] = v
	}
	return out, rows.Err()
}

File: frontend/src/stores/state.ts
import { acceptHMRUpdate, defineStore } from "pinia";
import { PutMany, PutSetting, Settings, Workspace } from "wailsjs/go/app/StateService";

// Durable state, in app.db rather than the webview's localStorage (which a
// reinstall, a new dev origin or a cleared cache silently empties).
//
// Per-workspace keys hold lenses, merges, facets and arrangements; settings
// hold global preferences. Values are JSON. Reads are synchronous against a
// hydrated copy; `hydrate(workspace)` must resolve before anything reads a
// workspace key, which is what the workspaces store awaits on select.
// Writes are debounced and flushed on close.

const FLUSH_MS = 300;

type Json = unknown;

export const useStateStore = defineStore("state", {
    state: () => ({
        workspace: "" as string,
        values: {} as Record<string, Json>,
        settings: {} as Record<string, Json>,
        hydrated: false,
        settingsLoaded: false,
    }),
    actions: {
        async hydrate(workspaceId: string) {
            await this.flush();
            if (this.workspace === workspaceId && this.hydrated) return;
            this.hydrated = false;
            this.workspace = workspaceId;
            let raw: Record<string, string> = {};
            try { raw = (await Workspace(workspaceId)) ?? {}; } catch { raw = {}; }
            // A stale response for a workspace no longer selected is dropped.
            if (this.workspace !== workspaceId) return;
            const values: Record<string, Json> = {};
            for (const [k, v] of Object.entries(raw)) values[k] = parse(v);
            this.values = values;
            this.hydrated = true;
        },
        async loadSettings() {
            if (this.settingsLoaded) return;
            let raw: Record<string, string> = {};
            try { raw = (await Settings()) ?? {}; } catch { raw = {}; }
            const out: Record<string, Json> = {};
            for (const [k, v] of Object.entries(raw)) out[k] = parse(v);
            this.settings = out;
            this.settingsLoaded = true;
        },
        get<T>(key: string, fallback: T): T {
            return (key in this.values ? (this.values[key] as T) : fallback);
        },
        /** Undefined or null deletes the key. */
        set(key: string, value: Json) {
            if (!this.workspace) return;
            const next = { ...this.values };
            if (value === undefined || value === null) delete next[key];
            else next[key] = value;
            this.values = next;
            pending.set(key, value === undefined || value === null ? "" : JSON.stringify(value));
            pendingWorkspace = this.workspace;
            schedule(() => this.flush());
        },
        setting<T>(key: string, fallback: T): T {
            return (key in this.settings ? (this.settings[key] as T) : fallback);
        },
        async setSetting(key: string, value: Json) {
            const next = { ...this.settings };
            if (value === undefined || value === null) delete next[key];
            else next[key] = value;
            this.settings = next;
            try { await PutSetting(key, value === undefined || value === null ? "" : JSON.stringify(value)); } catch { /* kept in memory */ }
        },
        async flush() {
            if (timer) { clearTimeout(timer); timer = null; }
            if (pending.size === 0 || !pendingWorkspace) return;
            const batch = Object.fromEntries(pending);
            const ws = pendingWorkspace;
            pending.clear();
            try { await PutMany(ws, batch); } catch { /* the next write retries the key */ }
        },
    },
});

const pending = new Map<string, string>();
let pendingWorkspace = "";
let timer: ReturnType<typeof setTimeout> | null = null;
function schedule(fn: () => void) {
    if (timer) clearTimeout(timer);
    timer = setTimeout(fn, FLUSH_MS);
}

function parse(v: string): Json {
    try { return JSON.parse(v); } catch { return v; }
}

// Closing the window must not lose the last 300 ms of changes.
if (typeof window !== "undefined") {
    window.addEventListener("beforeunload", () => {
        if (pending.size && pendingWorkspace) void PutMany(pendingWorkspace, Object.fromEntries(pending));
    });
}

if (import.meta.hot) import.meta.hot.accept(acceptHMRUpdate(useStateStore, import.meta.hot));
